### PR TITLE
Add tier logic

### PR DESCRIPTION
**Pull request by FionaQY**
Closes https://github.com/AY2425S1-CS2103T-T14-4/tp/issues/48

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 1]
```diff
@@ -44,9 +43,9 @@ public AddCommand parse(String args) throws ParseException {
         Phone phone = ParserUtil.parsePhone(argMultimap.getValue(PREFIX_PHONE).get());
         Email email = ParserUtil.parseEmail(argMultimap.getValue(PREFIX_EMAIL).get());
         Address address = ParserUtil.parseAddress(argMultimap.getValue(PREFIX_ADDRESS).get());
-        Set<Tag> tagList = ParserUtil.parseTags(argMultimap.getAllValues(PREFIX_TAG));
+        Tier tier = ParserUtil.parseTag(argMultimap.getValue(PREFIX_TAG).orElse("NA"));
```
Comment (ZShunRen) [inline src/main/java/seedu/address/logic/parser/AddCommandParser.java:46]: I like this default value. We just have to remember to document this down and possibly modify the UI to not show the label, if this is the text value for tier.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 2]
```diff
@@ -101,24 +101,24 @@ public static Email parseEmail(String email) throws ParseException {
...
+    public static Set<Tier> parseTags(Collection<String> tags) throws ParseException {
         requireNonNull(tags);
-        final Set<Tag> tagSet = new HashSet<>();
+        final Set<Tier> tierSet = new HashSet<>();
         for (String tagName : tags) {
-            tagSet.add(parseTag(tagName));
+            tierSet.add(parseTag(tagName));
         }
-        return tagSet;
+        return tierSet;
```
Comment (ZShunRen) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:122]: Perhaps we could remove this, as tiers are now a singular value and no longer a set.

File: src/main/java/seedu/address/model/tier/Tier.java [hunk 3]
```diff
@@ -56,7 +61,18 @@ public int hashCode() {
...
+
+    /**
+     * The Enums for what values Tier can take.
+     */
+    public enum TierEnum {
+        GOLD,
+        SILVER,
+        BRONZE,
+        REJECT,
+        NA
```
Comment (ZShunRen) [inline src/main/java/seedu/address/model/tier/Tier.java:75]: Nice job here, with the usage of enums!

File: src/main/java/seedu/address/storage/JsonAdaptedPerson.java [hunk 4]
```diff
@@ -30,24 +24,22 @@ class JsonAdaptedPerson {
...
+    private final JsonAdaptedTag tagged;
 
     /**
      * Constructs a {@code JsonAdaptedPerson} with the given person details.
      */
     @JsonCreator
     public JsonAdaptedPerson(@JsonProperty("name") String name, @JsonProperty("phone") String phone,
                              @JsonProperty("email") String email, @JsonProperty("address") String address,
-                             @JsonProperty("tagged") List<JsonAdaptedTag> tagged,
+                             @JsonProperty("tagged") JsonAdaptedTag tagged,
```
Comment (ZShunRen) [inline src/main/java/seedu/address/storage/JsonAdaptedPerson.java:35]: Perhaps we could replace the type here with just string and not `JsonAdaptedTag`.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 5]
```diff
@@ -42,33 +41,33 @@
...
 
 
         // multiple tags - all accepted
-        Person expectedPersonMultipleTags = new PersonBuilder(BOB).withTags(VALID_TAG_FRIEND, VALID_TAG_HUSBAND)
+        Person expectedPersonMultipleTags = new PersonBuilder(BOB).withTiers(VALID_TIER_GOLD)
                 .build();
         assertParseSuccess(parser,
-                NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_BOB + TAG_DESC_HUSBAND + TAG_DESC_FRIEND,
+                NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_BOB + TAG_DESC_REJECT + TAG_DESC_GOLD,
                 new AddCommand(expectedPersonMultipleTags));
```
Comment (ZShunRen) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:64]: I am not too sure that specifying multiple tiers should be acceptable.

File: src/main/java/seedu/address/model/person/Person.java [hunk 6]
```diff
@@ -101,13 +98,13 @@ public boolean equals(Object other) {
                 && phone.equals(otherPerson.phone)
                 && email.equals(otherPerson.email)
                 && address.equals(otherPerson.address)
-                && tags.equals(otherPerson.tags);
+                && tier.equals(otherPerson.tier);
```
Comment (ZShunRen) [inline src/main/java/seedu/address/model/person/Person.java:101]: Perhaps tiers not matching can still mean they are the same people, as perhaps they had a previous card tier already?

**Review by ZShunRen — APPROVED** [review]
LGTM for the most part. I just had a few questions about some of the tests, but overall good job at refactoring tiers from a set into a type that accepts a singular value based on an enum. However, I still think that there's remnant code.

**Comment by iamdiluxedbutcooler** [comment]
Resolve merge conflicts: mostly related to Person class, can take a look more on the constructor and tests

**Pull request merged**